### PR TITLE
[FLINK-22367][streaming] Reset syncSavepointId only if it is equal to…

### PR DESCRIPTION
**Pull request by akalash**
… checkpoint id from event

## What is the purpose of the change

*PR fixes the bug with resetting syncSavepointId at the wrong time which led to problem with terminateWithSavepointWithoutComplicationsShouldSucceedAndLeadJobToFinished*

## Brief change log

  - *Right now, syncSavepointId can be reset only if it is equal to the checkpoint id from event *

## Verifying this change
This change added tests and can be verified as follows:

  - *Added new test which checks that the abort before completing checkpoint  doesn't lead to the problem with finishing task. *

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java [hunk 1]
```diff
@@ -1264,6 +1265,42 @@ public void testCheckpointDeclinedOnClosedOperator() throws Throwable {
...
+                        .addInput(BasicTypeInfo.INT_TYPE_INFO);
+        StreamTaskMailboxTestHarness<Integer> harness =
+                builder.setupOutputForSingletonOperatorChain(
+                                new TestBoundedOneInputStreamOperator())
+                        .build();
+
+        // when: Receiving the abort notification of the previous checkpoint before the complete
+        // notification of the savepoint terminate.
+        harness.streamTask.notifyCheckpointAbortAsync(1);
+        harness.streamTask.notifyCheckpointCompleteAsync(2);
```
Comment (dawidwys) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java:1292]: How can this happen? You cannot complete a checkpoint that has not been triggered yet. The order of `notifyCheckpointCompleteAsync` and `triggerCheckpointOnBarrier` is wrong here.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java [hunk 2]
```diff
@@ -450,7 +450,10 @@ private void resetSynchronousSavepointId(long id, boolean succeeded) {
             activeSyncSavepointId = null;
             operatorChain.setIgnoreEndOfInput(false);
         }
-        syncSavepointId = null;
+
+        if (syncSavepointId != null && syncSavepointId == id) {
```
Comment (dawidwys) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:454]: How about we change it to?: 
```suggestion
        if (syncSavepointId != null && syncSavepointId <= id) {
```

**Review by dawidwys — COMMENTED** [review]
I think your investigation is valid. Have you checked if it is a problem for earlier versions that we support? E.g. 1.12.x?

AFAICT it should not be a problem for master. It should've been fixed there. One change I'd do is to reset it for any newer checkpoint than that of the savepoint. Just to stay on the safe side. WDYT?

It should not happen, as there should be no checkpoints triggered from the checkpoint coordinator after a sync savepoint (it should either stop the job or trigger a failover).

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java (hunk 1 above)
Comment (akalash) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java:1292]: I took this order from the test `terminateWithSavepointWithoutComplicationsShouldSucceedAndLeadJobToFinished`. Usually, it indeed happens in normal order, but sometimes I see that the abort happens before the complete.

Comment (dawidwys) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java:1292]: I meant the order of `notifyCheckpointCompleteAsync` and `triggerCheckpointOnBarrier`.

The barrier must arrive first before we can say the checkpoint completed. It's not possible to receive a `notifyCheckpointCompleteAsync` before we've seen a barrier for the checkpoint.

In your testcase you're first sending the `notifyCheckpointCompleteAsync(2)` and after that you're sending a barrier: `triggerCheckpointOnBarrier(...(2)...)`

**Comment by akalash** [comment]
It is also valid for at least release-1.12 -  https://github.com/apache/flink/pull/17104

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java (hunk 1 above)
Comment (akalash) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java:1292]: `notifyCheckpointCompleteAsync` and `notifyCheckpointAbortAsync` actually do nothing but add the task to mailbox. When `triggerCheckpointOnBarrier` is called it first of all does its own job and only then it directly execute all waiting mails until the checkpoint is complete. So in fact the `triggerCheckpointOnBarrier` is always first.

Comment (dawidwys) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java:1292]: Still this is based on the internal knowledge of the component. It is still true that those methods can not be called in that order in the production code. I find the current order both confusing and prone to changes in the implementation. I find it also distracting from the real scenario that's being tested here.

I might be wrong, but I truly believe I am not the only one that would be confused by this order.

Comment (akalash) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java:1292]: Do I understand correctly, that you are talking about order in the test? Do you mean that we should explicitly call `triggerCheckpointOnBarrier` first? If so please, check the my last commit I a little rewrote the test

Comment (dawidwys) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java]: Yes, I am talking about the order in the test.

Why wasn't it necessary in the previous version to call the `triggerCheckpointOnBarrier` from within the executor? 

Do we need to call the `notifyCheckpointCompleteAsync` and notifyCheckpointAbortAsync` from the executor? I think those methods are called via RPC in the production code and are not executed from the executor thread...

Comment (dawidwys) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java]: Ok, I checked it myself. If we execute it from the test thread it gets stuck in the synchronous savepoint loop. 

I think executing `triggerCheckpointOnBarrier` from the `MailboxExecutor` is fine here. I would not do it for the `*Async` methods.

Comment (akalash) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java:1303]: ok, I have removed the mailbox Executor for async calls. Initially, I added it that only for keeping the one style but it indeed didn't have any real function.

**Review by dawidwys — APPROVED** [review]
LGTM, let's wait for Azure

**Pull request closed without merging**